onnxruntime/core/providers/cpu/quantization/quantize_linear.cc

```diff
@@ -244,66 +265,218 @@ ONNX_CPU_OPERATOR_TYPED_MS_KERNEL(
-        for (size_t bs = 0; bs < static_cast<size_t>(quant_block_size); bs++) {
+struct DequantizeLinearApply<T, OutT, false> {
+  /**
+   * @brief Calculate per-tensor/layer or per-axis quantization of DequantizeLinear on the
+   *        flattened tensors.
+   * @param[in]    M                      size of dimensions before the quantize axis
+   * @param[in]    K                      dimension on the quantize axis
+   * @param[in]    N                      size of dimensions after the quantize axis
+   * @param[in]    input                  1D array of flattened [D0, ..., Di, ..., Dn]
+   * @param[in]    scale                  scalar for per-tensor/layer quantization and 1D array [Di]
+   *                                      for per-axis quantization. i is the quantize axis.
+   * @param[out]   output                 same shape as input
+   * @param[in]    zero_point             same shape as scale
+   */
+  void op(int64_t M, int64_t K, int64_t N, const T* input,
+          const OutT* scale, OutT* output, const T* zero_point) {
+    for (size_t m = 0; m < static_cast<size_t>(M); m++) {
+      for (size_t k = 0; k < static_cast<size_t>(K); k++) {
+        auto zp = zero_point ? static_cast<int32_t>(zero_point[k]) : 0;
+        auto sc = static_cast<float>(scale[k]);
+        for (size_t n = 0; n < static_cast<size_t>(N); n++) {
           *output++ = static_cast<OutT>(static_cast<float>(static_cast<int32_t>(*input++) - zp) * sc);
         }
       }
     }
   }
+
+  /**
+   * @brief Calculate blocked quantization of DequantizeLinear on the flattened tensors.
+   *        TODO(fajin): add mlas kernel to utilize multithreading, refer MlasDequantizeBlockwise.
+   * @param[in]    M                      size of dimensions before the quantize axis
+   * @param[in]    K                      dimension of the quantize axis
+   * @param[in]    N                      size of dimensions after the quantize axis
+   * @param[in]    quant_block_size       quantize block size along the quantize axis
+   * @param[in]    input                  1D array of flattened [D0, ..., Di, ..., Dn]
+   * @param[in]    scale                  1D array of flattened [D0, ..., ceil(Di/quant_block_size), ..., Dn].
+   *                                      i is the quantize axis.
+   * @param[out]   output                 same shape as input
+   * @param[in]    zero_point             same shape as scale
+   */
+  void op(int64_t M, int64_t K, int64_t N, int64_t quant_block_size,
+          const T* input, const OutT* scale, OutT* output, const T* zero_point) {
+    auto qbsiz = static_cast<size_t>(quant_block_size);
+    auto bdim = static_cast<size_t>(K);
+    auto pbsiz = static_cast<size_t>(N);
+
+    if (zero_point) {
+      for (size_t m = 0; m < static_cast<size_t>(M); m++) {
+        for (size_t bd = 0; bd < bdim; bd += qbsiz) {
+          size_t bd_q_stop = bd + qbsiz > bdim ? bdim : bd + qbsiz;
+
+          for (size_t qb = bd; qb < bd_q_stop; ++qb) {
```

Comment:
>for (size_t qb = bd; qb < bd_q_stop; ++qb) { [](http://example.com/codeflow?start=10&length=44)

nit:
for(size_t qb = 0; qb < min(qbsize, bdim - bd); qb++)

Reply:
updated